feat: test pg_upgrade compatibility with older extension versions

jfroche opened the pull request:
Add test to verify that all extension versions from PostgreSQL 15 can successfully upgrade to PostgreSQL 17 using pg_upgrade.

The test now validates:

- Each PG 15 extension version can be upgraded to PG 17
- Extension update scripts are properly generated during upgrade
- Version handling works correctly with and without update scripts
- Final extension versions match expected values after upgrade

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Unified PostgreSQL-version-aware packaging that produces SQL-only migration artifacts for older Postgres releases and includes them in final builds.

* **Tests**
  * Broadened pg_upgrade 15→17 test coverage across many extensions; added a dedicated pg_stat_monitor test and standardized per-version upgrade flows. Test exports updated to include pg_partman and postgis.

* **Bug Fixes**
  * Test cleanup now drops extensions with CASCADE to remove dependent objects.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

jfroche committed:
feat: test pg_upgrade compatibility with older extension versions

Adds pg_upgrade compatibility testing in default.nix to verify that extension
versions from PostgreSQL 15 can successfully upgrade to PostgreSQL 17.

The test validates:
- Each PG 15 extension version can be upgraded to PG 17
- Extension update scripts are properly generated during upgrade
- Version handling works correctly with and without update scripts
- Final extension versions match expected values after upgrade

jfroche committed:
fix: include unsupported extension versions as SQL-only for pg_upgrade compatibility

When upgrading PostgreSQL versions, pg_upgrade needs access to old extension
SQL files to migrate from. This adds unsupported pg_graphql and pg_jsonschema
versions (those not compilable with current PostgreSQL) as SQL-only packages,
ensuring migration paths exist when upgrading from older PostgreSQL versions.

samrose reviewed (changes requested):
Looks great! Will just need to test before we merge it

yvan-sraka committed:
fix(pg_stat_monitor): add custom pg_upgrade compatibility test

Implements custom test for pg_stat_monitor with special handling for version
2.1 schema differences between PostgreSQL 15 and 17. Version 1.0 is excluded
from pg_upgrade tests as it uses .sql.in template files that reference
MODULE_PATHNAME without proper processing for the target version.

Version 2.1 has different schemas on PostgreSQL 15 vs 17 despite sharing the
same version number. During pg_upgrade, the extension version remains 2.1
without schema migration since no update script is generated. The custom test
skips pg_regress validation after pg_upgrade when no update script is
generated, since schema mismatch is expected behavior.

Removes pg_stat_monitor from default extension test list and uses
pkgs.testers.runNixOSTest with consistent pkgs.pkgsLinux references for
cross-platform compatibility.

yvan-sraka committed:
fix(pg_jsonschema): remove debug find command from build output

yvan-sraka committed:
fix(pg_graphql): use string comparison for pgVersion

pgVersion is a string, so comparison should be string-to-string
instead of string-to-integer.

yvan-sraka commented on nix/ext/pg_graphql/default.nix:
```diff
@@ -119,19 +119,64 @@ let
+          fi
+        done > $out/share/postgresql/extension/${pname}--${version}.sql
+      '';
+      meta = with lib; {
+        description = "GraphQL support for PostreSQL";
+        homepage = "https://github.com/supabase/${pname}";
+        license = licenses.postgresql;
+        inherit (postgresql.meta) platforms;
+      };
```
Funny typo, I suggest we fix it in another PR:
```
postgres on git develop
> rg Postre
README.md
201:| [pg_graphql](https://github.com/supabase/pg_graphql/archive/v1.5.11.tar.gz) | [1.5.11](https://github.com/supabase/pg_graphql/archive/v1.5.11.tar.gz) | GraphQL support for PostreSQL |
235:| [pg_graphql](https://github.com/supabase/pg_graphql/archive/v1.5.11.tar.gz) | [1.5.11](https://github.com/supabase/pg_graphql/archive/v1.5.11.tar.gz) | GraphQL support for PostreSQL |
268:| [pg_graphql](https://github.com/supabase/pg_graphql/archive/v1.5.11.tar.gz) | [1.5.11](https://github.com/supabase/pg_graphql/archive/v1.5.11.tar.gz) | GraphQL support for PostreSQL |

nix/ext/wrappers/default.nix
161:          description = "Various Foreign Data Wrappers (FDWs) for PostreSQL";

nix/ext/pg_graphql/default.nix
99:          description = "GraphQL support for PostreSQL";
```

yvan-sraka replied:
If you can open a new PR that fix it, that would be even better ;)

yvan-sraka committed:
fix(tests): handle list-type shared_preload_libraries and pgsodium compatibility

Converts list-type shared_preload_libraries to comma-separated strings for
pg_upgrade --old-options and --new-options flags using lib.concatStringsSep.
Fixes Python f-string interpolation for psql_17 and orioledb_17 variables in
test scripts.

Skips upgrade path tests for pgsodium where supautils before-create.sql hook
enforces that only version 3.1.8 can be installed, preventing creation of
older versions needed for upgrade testing.

jfroche committed:
feat: limit nix-eval-jobs max memory to 3 GiB per worker

jfroche committed:
feat: log nix-eval-jobs command to stderr
